numpy/core/src/multiarray/ctors.c

```diff
@@ -1347,6 +1349,12 @@ _array_from_array_like(PyObject *op,
      *      this should be changed!
      */
     if (!writeable && tmp == Py_NotImplemented) {
+        PyObject* array_meth = PyArray_LookupSpecial_OnInstance(op, "__array__");
+        int has_get = array_meth && PyType_Check(op) && PyObject_HasAttrString(array_meth, "__get__");
+        if (array_meth != NULL && !has_get && do_copy) {
+            PyErr_SetString(PyExc_ValueError, "Calling __array__ in never copy mode is not allowed.");
+            return NULL;
+        }
```

Comment:
A bit annoying, but I am wondering if we should introduce a new `PyArray_FromArrayAttr_int` with `PyObject *context` replaced with this flag (not public for now of course).

That way we can avoid the dance here and are also a bit optimistic about a future where we might pass an argument to `__array__` instead of refusing to call it.